ecosystem/node-checker/src/evaluators/direct/node_identity.rs

```diff
@@ -0,0 +1,153 @@
+// Copyright (c) Aptos
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{
+    configuration::{EvaluatorArgs, NodeAddress},
+    evaluator::{EvaluationResult, Evaluator},
+};
+use anyhow::{anyhow, format_err, Result};
+use aptos_config::config::RoleType;
+use aptos_sdk::types::chain_id::ChainId;
+use clap::Parser;
+use poem_openapi::Object as PoemObject;
+use serde::{Deserialize, Serialize};
+use std::{collections::HashMap, str::FromStr};
+use thiserror::Error as ThisError;
+
+use super::DirectEvaluatorInput;
+
+pub const CATEGORY: &str = "node_identity";
+
+/// This function hits the `/` endpoint of the API and returns the chain ID
+/// and role type, extracted from the IndexResponse.
+pub async fn get_node_identity(node_address: &NodeAddress) -> Result<(ChainId, RoleType)> {
+    let mut url = node_address.url.clone();
+    url.set_port(Some(node_address.api_port))
+        .map_err(|_| format_err!("Failed to set port for URL"))?;
+
+    let response = reqwest::get(url)
+        .await
+        .map_err(|e| format_err!("Failed to get node identity {}", e))?;
+    let response_body = response
+        .text()
+        .await
+        .map_err(|e| format_err!("Failed to get body of node identity response {}", e))?;
+
+    let data: HashMap<String, serde_json::Value> =
+        serde_json::from_str(&response_body).map_err(|e| {
+            format_err!(
+                "Failed to process response body as valid JSON with string key/values {}",
+                e
+            )
+        })?;
+
+    let chain_id_raw: u8 = data
+        .get("chain_id")
+        .ok_or_else(|| format_err!("Failed to get chain_id from node identity"))?
+        .as_u64()
+        .ok_or_else(|| anyhow!("Failed to read chain ID from node identity as u8"))?
+        as u8;
+    let chain_id = ChainId::new(chain_id_raw);
+
+    let role_type_raw = data
+        .get("node_role")
+        .ok_or_else(|| format_err!("Failed to get node_role from node identity"))?
+        .as_str()
+        .ok_or_else(|| anyhow!("Failed to read node_role from node identity as str"))?;
+    let role_type = RoleType::from_str(role_type_raw)
+        .map_err(|e| format_err!("Failed to parse node_role {}", e))?;
+
+    Ok((chain_id, role_type))
+}
+
+#[derive(Debug, ThisError)]
+pub enum NodeIdentityEvaluatorError {
+    #[error("Failed to get node identity from node: {0}")]
+    GetNodeIdentityError(anyhow::Error),
+}
+
+// TODO: Consider taking chain_id and role_type here instead.
```

Comment:
Yeah, I guess then we'd actually be able to write some unit tests for this as we wouldn't need to ping endpoint urls in the tests?